src/perf_counters.cc

```diff
@@ -26,9 +26,121 @@
+  SinglePMURegistry(SinglePMURegistry&&) = default;
+  SinglePMURegistry(const SinglePMURegistry&) = delete;
+  SinglePMURegistry& operator=(SinglePMURegistry&&) noexcept;
+  SinglePMURegistry& operator=(const SinglePMURegistry&) = delete;
+
+  SinglePMURegistry(pfm_pmu_t pmu_id)
+      : pmu_id_(pmu_id), available_counters_(0), available_fixed_counters_(0) {
+    {
+      pfm_pmu_info_t pmu_info{};
+      const auto pfm_pmu = pfm_get_pmu_info(pmu_id, &pmu_info);
+
+      if (pfm_pmu != PFM_SUCCESS) {
+        GetErrorLogInstance() << "Unknown pmu: " << pmu_id << "\n";
+        return;
+      }
+
+      name_ = pmu_info.name;
+      desc_ = pmu_info.desc;
+      available_counters_ = pmu_info.num_cntrs;
+      available_fixed_counters_ = pmu_info.num_fixed_cntrs;
+
+      BM_VLOG(1) << "PMU: " << pmu_id << " " << name_ << " " << desc_ << "\n";
+      BM_VLOG(1) << "     counters: " << available_counters_ << " fixed: " << available_fixed_counters_ << "\n";
+    }
+  }
+
+  const char* name() const { return name_; }
+
+  bool AddCounter(int event_id) {
+    pfm_event_info_t info{};
+    const auto pfm_event_info =
+        pfm_get_event_info(event_id, PFM_OS_PERF_EVENT, &info);
+
+    if (pfm_event_info != PFM_SUCCESS) {
+      GetErrorLogInstance() << "Unknown event id: " << event_id << "\n";
+      return false;
+    }
+
+    assert(info.pmu == pmu_id_);
+
+    if (counter_ids_.find(event_id) != counter_ids_.end()) return true;
+
+    if (counter_ids_.size() >= available_counters_ - 1) {
+      GetErrorLogInstance() << "Maximal number of counters for PMU " << name_
+                            << " (" << available_counters_ << ") reached.\n";
+      return false;
+    }
+
+    counter_ids_.emplace(event_id, info.code);
+
+    BM_VLOG(2) << "Registered counter: " << event_id << " (" << info.name << " - " << info.desc
+               << ") in pmu " << name_ << " (" << counter_ids_.size() << "/" << available_counters_ << "\n";
+
+    return true;
+  }
+
+ private:
+  pfm_pmu_t pmu_id_;
+  const char* name_;
+  const char* desc_;
+  std::unordered_map<int, uint64_t> counter_ids_;
+  std::unordered_map<int, uint64_t> fixed_counter_ids_;
+  uint64_t available_counters_;
+  uint64_t available_fixed_counters_;
+};
+
+class PMURegistry {
+ public:
+  ~PMURegistry() = default;
+  PMURegistry(PMURegistry&&) = default;
+  PMURegistry(const PMURegistry&) = delete;
+  PMURegistry& operator=(PMURegistry&&) noexcept;
+  PMURegistry& operator=(const PMURegistry&) = delete;
+  PMURegistry() {}
+
+  bool EnlistCounter(const std::basic_string<char>& name, struct perf_event_attr &attr_base) {
+    attr_base.size = sizeof(attr_base);
+    pfm_perf_encode_arg_t encoding{};
+    encoding.attr = &attr_base;
+
+    const auto pfm_get = pfm_get_os_event_encoding(
+        name.c_str(), PFM_PLM3, PFM_OS_PERF_EVENT, &encoding);
+    if (pfm_get != PFM_SUCCESS) {
+      GetErrorLogInstance() << "Unknown counter name: " << name << "\n";
+      return false;
```

Comment:
just checking design decision: given the perf counters are requested by users would it not be better to hard-fail here with an assert?

Reply:
Agree 100% percent, this is why I'm always returning `false` from the internal api and later failing with:

```cpp
    if (!ok) {
      GetErrorLogInstance() << "Failed to register counter: " << name << "\n";
      return NoCounters();
    }
```

Reply:
ok. i'd also be ok with us just asserting/CHECKing here, given it's all one library. up to you.